Adds telemetry

joladev opened the pull request:
Adds span style events for pull request and ack, including tests and documentation.

Happy to take suggestions on naming. Potentially the existing demand is not something we should emit here? Added it cause it'd find it personally useful but happy to remove if it doesn't make sense.

It was a bit hard to do this because the config passed to ack doesn't come from the same place at all as the config passed to receive_messages, the first being just validated producer opts put in persistent term and not as you'd expect, the ack opts validated in the configure callback. And the second being the full config passed to the producer, including the broadway config. The test config doesn't really match the config when running for real, eg in the tests the `receive_messages` tests get the same config as the `ack` tests

josevalim commented on lib/broadway_cloud_pub_sub/pull_client.ex:
```diff
@@ -36,10 +36,23 @@ defmodule BroadwayCloudPubSub.PullClient do
+          |> handle_response(:receive_messages)
+          |> wrap_received_messages(ack_builder)
+
+        {result, %{topology_name: config.broadway[:name]}}
```
Typically the stop metadata should be a superset of the start one. :)

joladev replied:
Sure, will add

joladev replied:
done

josevalim commented:
I think we should at least align the metadata naming with SQS. WDYT? https://hexdocs.pm/broadway_sqs/BroadwaySQS.Producer.html#module-telemetry

Aligning the event name would be good but not required. Thoughts?

joladev commented:
> I think we should at least align the metadata naming with SQS. WDYT? https://hexdocs.pm/broadway_sqs/BroadwaySQS.Producer.html#module-telemetry
> 
> Aligning the event name would be good but not required. Thoughts?

Ah of course, switching

joladev commented:
Although looking at that, it only has `demand` (is that total demand or how many messages were requested?). In this case what's useful to expose is what the PullClient is setting as `maxMessages`. So maybe naming it `max_messages`?

joladev commented:
Exposing messages seems unnecessary, since Broadway already does that

joladev commented:
Made a suggested change, renaming `total_demand` -> `demand`, `requested_messages` -> `max_messages`. Added them to stop event.

Also renamed `pull_request` event name to `receive_messages`.

josevalim commented on lib/broadway_cloud_pub_sub/producer.ex:
```diff
@@ -153,7 +175,9 @@ defmodule BroadwayCloudPubSub.Producer do
     {specs, opts} = prepare_to_connect(broadway_opts[:name], client, opts)
 
-    :persistent_term.put(ack_ref, Map.new(opts))
+    ack_opts = Keyword.put(opts, :topology_name, broadway_opts[:name])
+
+    :persistent_term.put(ack_ref, Map.new(ack_opts))
```
The `ack_ref` itself is the topology name, so I think we can keep the previous one:

```suggestion
    :persistent_term.put(ack_ref, Map.new(opts))
```

josevalim replied:
Ignore. :)

josevalim commented on lib/broadway_cloud_pub_sub/pull_client.ex:
```diff
@@ -36,10 +36,24 @@ defmodule BroadwayCloudPubSub.PullClient do
+      %{
+        max_messages: max_messages,
+        demand: demand,
+        topology_name: config.broadway[:name]
```
To align with SQS:

```suggestion
        name: config.broadway[:name]
```

joladev replied:
Shouldn't this be `topology_name` to match Broadway's own telemetry? `name` seems confusing since it's the name of the topology but called `name` but emitted by the `PullClient`

josevalim commented on lib/broadway_cloud_pub_sub/pull_client.ex:
```diff
@@ -56,9 +70,18 @@ defmodule BroadwayCloudPubSub.PullClient do
+          |> execute(:acknowledge, %{"ackIds" => ack_ids})
+          |> handle_response(:acknowledge)
+
+        {result, %{topology_name: config.topology_name}}
```
```suggestion
        {result, %{name: config.topology_name}}
```

josevalim commented on test/broadway_cloud_pub_sub/pull_client_test.exs:
```diff
@@ -247,6 +280,39 @@ defmodule BroadwayCloudPubSub.PullClientTest do
+      PullClient.acknowledge(["1", "2", "3"], opts)
+
+      assert_received {:start, metadata}
+      assert metadata.topology_name == Broadway3
```
```suggestion
      assert metadata.name == Broadway3
```

josevalim commented on test/broadway_cloud_pub_sub/pull_client_test.exs:
```diff
@@ -247,6 +280,39 @@ defmodule BroadwayCloudPubSub.PullClientTest do
+      assert metadata.topology_name == Broadway3
+      assert_received {:stop, measurements, metadata}
+      assert is_integer(measurements.duration)
+      assert metadata.topology_name == Broadway3
```
```suggestion
      assert metadata.name == Broadway3
```

josevalim commented on lib/broadway_cloud_pub_sub/pull_client.ex:
```diff
@@ -36,10 +36,24 @@ defmodule BroadwayCloudPubSub.PullClient do
+          |> handle_response(:receive_messages)
+          |> wrap_received_messages(ack_builder)
+
+        {result,
+         %{name: config.broadway[:name], max_messages: max_messages, demand: demand}}
```
```suggestion
        {result, %{name: config.broadway[:name], max_messages: max_messages, demand: demand}}
```

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
